File: dynamic_programming/05_climb_stairs_min_cost/climb_stairs_min_cost.py
from typing import List


class Solution:
    def min_cost_climbing_stairs(self, cost: List[int]) -> int:
        if not self._check_constraints(cost):
            raise ValueError('Constraints violated')
        n = len(cost)
        # Break optimal solution to the problem into optimal solutions to subproblems
        # 1st method: backward memoization
        memo: List[int] = [-1] * (n + 1)
        return self._backward_memo(0, n, memo, cost)

        # 2nd method: forward dynamic programming
        dp: List[int] = [0] * (n + 1)
        dp[0] = cost[1]
        dp[1] = cost[2]
        for i in range(2, n + 1):
            dp[i] = cost[i] if i < n else 0
            dp[i] += min(dp[i - 1], dp[i - 2])
        return dp[n]

    def _check_constraints(self, cost: List[int]) -> bool:
        if len(cost) < 2 or len(cost) > 1000:
            return False
        for c in cost:
            if c < 0 or c > 999:
                return False
        return True

    def _backward_memo(self, i: int, n: int, memo: List[int], cost: List[int]) -> int:
        if i > n:  # overshoot
            return 0
        if memo[i] > -1:  # computed before
            return memo[i]
        memo[i] = cost[i - 1] if i > 0 else 0
        memo[i] += min(self._backward_memo(i + 1, n, memo, cost), self._backward_memo(i + 2, n, memo, cost))  # forward step of size 1 and 2
        return memo[i]

File: dynamic_programming/05_climb_stairs_min_cost/climb_stairs_min_cost.cpp
#include <iostream>
#include <stdexcept>
#include <cmath>
#include <cstdint>
#include <vector>
#include <algorithm>

using namespace std;


class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        if (!checkConstraints(cost)) {
            throw std::runtime_error("Constraints violated");
        }
        int n = cost.size();
        // Break optimal solution to the problem into optimal solutions to subproblems
        // 1st method: backward memoization
        // vector<int> memo(n + 1, -1);
        // return backward_memo(0, n, memo, cost);
        // 2nd method: forward dynamic programming 
        vector<int> dp(n + 1, 0);
        dp[0] = cost[0];
        dp[1] = cost[1];
        for (int i = 2; i <= n; i++) {
            dp[i] = i < n ? cost[i] : 0;
            dp[i] += min(dp[i - 1], dp[i - 2]);
        }
        return dp[n];
    }

private:
    bool checkConstraints(vector<int>& cost) {
        if (cost.size() < 2 || cost.size() > 1000) return false;
        for (int c : cost) {
            if (c < 0 || c > 999) return false;
        }
        return true;
    }
    int backward_memo(int i, int n, vector<int>& memo, vector<int> cost) {
        if (i > n) return 0; // overshoot
        if (memo[i] > -1) return memo[i]; // computed before
        memo[i] = i > 0 ? cost[i - 1] : 0;
        memo[i] += min(backward_memo(i + 1, n, memo, cost), backward_memo(i + 2, n, memo, cost)); // forward step of size 1 and 2
        return memo[i];
    }
};
